### DISCUSSION TITLE
Data: create only one line per participant

### DISCUSSION THREAD

**Discussion by kernel52** [jspsych/jsPsych#1972] [Q&A] [answered]
Is it possible to change the way data is saved throughout the timeline? 

What I try to achieve is to have the collected answers in one line per participant, rather than one line per trial. 

I was looking either for a way to suppress some of the trials from writing in the data object or a way to "merge" the trial data. 

Of course, this would be even more useful if I could do it inside cognition.run (where I'm afraid that is not possible at all to remove some of the info that it collects, like user_agent, browser_version, etc.)

any help would be much appreciated.

**Comment by jodeleeuw** [comment, marked as answer]
Hi @kernel52,

One-line-per-trial is pretty built in to the core of how jsPsych manages the data structure. If you only want to save a few values per participant then you could write a custom function to run at the end of the experiment that created a single array or object of data and send this off to some server. 

If you are working in cognition.run then you won't have enough control over the backend to do this. You could, in theory, use the `call-function` plugin to perform the data summarization and return the object containing the data. Then all of the data for each participant would be stored in the trial associated with the `call-function` plugin running. You could then filter the CSV file to just those rows.

But, there are simpler ways to do this outside of JavaScript where you can take advantage of data wrangling pipelines. My recommendation would be to focus on post-processing the data you get to wrangle it into your preferred format.

**Reply by kernel52** [reply]
Thank you for the reply Josh, 

I do see the idea behind the one-line-per-trial, and I agree that post-process has way more benefits. I just can't teach an old dog new tricks :?
I'm developing solutions for others (not a researcher myself) and they have a difficulty in adapting to the novelty, yet they love the facility your library is offering ;) 

What I was thinking as a solution, was to be able to suppress the output for some plugins, like the preload for a example that sometimes seems to be a line without much information for the end user. Would you think of a suppress option in the plugins for general? Kind of a flag to not print anything in the data should the user decides to? 

I tried to use the `save_trial_parameters` but I can only limit a few of the output fields, I also tried to delete data during the `on_finish` function of the plugins but is not a very "pretty" solution, let alone that in Cognition it doesn't even have a result, the data is collected no matter what I do  :D -but that's a different problem for a different person. 

In any case, thank you for the answer, you already gave me a solution to think of :) 

Best,
Nikolas

**Reply by jodeleeuw** [reply]
> Would you think of a suppress option in the plugins for general? Kind of a flag to not print anything in the data should the user decides to?

You can sort of do this already by using the `filter()` or `filterCustom()` functions in `jsPsych.data`. This can give you a data collection with rows omitted which you can then convert using `.csv()` or `.json()`.

We do have a simple flag on the roadmap, but it's not a high priority at the moment given the alternative solution that's available: https://github.com/jspsych/jsPsych/issues/687.

**Reply by kernel52** [reply]
Thanks! I will explore this path with `filter()` or `filterCustom()`

Indeed, is a nice-to-have feature but not a critical issue to be solved.

**Reply by kernel52** [reply]
In case someone tries to accomplish this on cognition.run, 
I managed to find a solution to my problem. 

on each of the trials I don't want to receive any data, I added the following 
```
 on_finish: function(data) {
     raw_data = jsPsych.data.get().values(); //https://www.jspsych.org/core_library/jspsych-data/#values
     raw_data.pop(); //https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop
   },
```

I was reading one comment of Javi that he's using the on_finish. 
At first I thought he used the function on the timeline, but it seems he's playing on a trial level (smart, I must admint). 

anyway, until something more elegant is found... :)
